File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return helper(root, p, q);
    }
    
    private TreeNode helper(TreeNode node, TreeNode p, TreeNode q) {
        if ((p.val < node.val && q.val > node.val) ||
            (p.val > node.val && q.val < node.val) ||
            (p.val == node.val) ||
            (q.val == node.val)) {
            
            return node;
        }
        
        if (p.val < node.val && q.val < node.val) {
            // search left
            return helper(node.left, p, q);
        } else {
            // search right
            return helper(node.right, p, q);
        }
    }
}

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # approach: 3 cases
        # if p and q < root, go left
        # if p and q > root, go right
        # else return root
        
        res = None
        
        def helper(node):
            nonlocal res
            nonlocal p
            nonlocal q
            
            if not node:
                return
            
            if p.val < node.val and q.val < node.val:
                helper(node.left)
            elif p.val > node.val and q.val > node.val:
                helper(node.right)
            else:
                res = node
        
        helper(root)
        return res
                